Mathlib/Tactic/CC/Addition.lean

```diff
@@ -0,0 +1,2061 @@
+    return tEqs
+  else if tr == sr then
+    let some tre := tr.toExpr | failure
+    DelayedExpr.ofExpr <$> mkEqRefl tre
+  else
+    let .apps op _ := tr | failure
+    let some re := r.toExpr | failure
+    let opr := op.app re
+    let some te := t.toExpr | failure
+    let rt := opr.app te
+    let some se := s.toExpr | failure
+    let rs := mkApp2 op re se
+    let rtEqrs := DelayedExpr.congrArg opr tEqs
+    let some tre := tr.toExpr | failure
+    let trEqrt ← mkACProof tre rt
+    let some sre := sr.toExpr | failure
+    let rsEqsr ← mkACProof rs sre
+    return .eqTrans (.eqTrans trEqrt rtEqrs) rsEqsr
+
+/-- Given `ra := a*r` `sb := b*s` `ts := t*s` `tr := t*r` `tsEqa : t*s = a` `trEqb : t*r = b`,
+    return a proof for `ra = sb`.
+
+    We use `a*b` to denote an AC application. That is, `(a*b)*(c*a)` is the term `a*a*b*c`. -/
+def mkACSuperposeProof (ra sb a b r s ts tr : ACApps) (tsEqa trEqb : DelayedExpr) :
+    MetaM DelayedExpr := do
+  let .apps _ _ := tr | failure
+  let .apps op _ := ts | failure
+  let tsrEqar := DelayedExpr.congrFun (.congrArg op tsEqa) r
+  let trsEqbs := DelayedExpr.congrFun (.congrArg op trEqb) s
+  let some tse := ts.toExpr | failure
+  let some re := r.toExpr | failure
+  let tsr := mkApp2 op tse re
+  let some tre := tr.toExpr | failure
+  let some se := s.toExpr | failure
+  let trs := mkApp2 op tre se
+  let tsrEqtrs ← mkACProof tsr trs
+  let some ae := a.toExpr | failure
+  let ar := mkApp2 op ae re
+  let some be := b.toExpr | failure
+  let bs := mkApp2 op be se
+  let some rae := ra.toExpr | failure
+  let raEqar ← mkACProof rae ar
+  let some sbe := sb.toExpr | failure
+  let bsEqsb ← mkACProof bs sbe
+  return .eqTrans raEqar (.eqTrans (.eqSymm tsrEqar) (.eqTrans tsrEqtrs (.eqTrans trsEqbs bsEqsb)))
+
+/-- Given `e := lhs * r` and `H : lhs = rhs`, return `rhs * r` and the proof of `e = rhs * r`. -/
+def simplifyACCore (e lhs rhs : ACApps) (H : DelayedExpr) :
+    CCM (ACApps × DelayedExpr) := do
+  guard (lhs.isSubset e)
+  if e == lhs then
+    return (rhs, H)
+  else
+    let .apps op _ := e | failure
+    let newArgs := e.diff lhs
+    let r : ACApps := if newArgs.isEmpty then default else .mkApps op newArgs
+    let newArgs := ACApps.append op rhs newArgs
+    let newE := ACApps.mkApps op newArgs
+    let some true := (← get).opInfo.find? op | failure
+    let newPr ← mkACSimpProof e lhs rhs r newE H
+    return (newE, newPr)
+
+/-- The single step of `simplifyAC`. -/
+def simplifyACStep (e : ACApps) : CCM (Option (ACApps × DelayedExpr)) := do
+  if let .apps _ args := e then
+    for h : i in [:args.size] do
+      if i == 0 || (args[i]'h.2) != (args[i - 1]'(Nat.lt_of_le_of_lt (i.sub_le 1) h.2)) then
+        let some ae := (← get).acEntries.find? (args[i]'h.2) | failure
+        let occs := ae.RLHSOccs
+        let mut Rlhs? : Option ACApps := none
+        for Rlhs in occs do
+          if Rlhs?.isNone then
+            if Rlhs.isSubset e then
+              Rlhs? := some Rlhs
```

Comment:
Instead of checking for `Rlhs?` being `none`, we can also `break`, right?
```suggestion
          if Rlhs.isSubset e then
            Rlhs? := some Rlhs
            break
```